src/Avalonia.Documents/Documents/InlineCollection.cs

```diff
@@ -0,0 +1,269 @@
+    /// of Paragraph, Span and TextBlock elements.
+    /// </summary>
+    [ContentWrapper(typeof(Run))]
+    [ContentWrapper(typeof(InlineUIContainer))]
+    [WhitespaceSignificantCollection]
+    public class InlineCollection : TextElementCollection<Inline>
+    {
+        //-------------------------------------------------------------------
+        //
+        //  Constructors
+        //
+        //-------------------------------------------------------------------
+
+        #region Constructors
+
+        // Constructor is internal. We allow InlineCollection creation only from inside owning elements such as TextBlock or TextElement.
+        // Note that when a SiblingInlines collection is created for an Inline, the owner of collection is that member Inline object.
+        // Flag isOwnerParent indicates whether owner is a parent or a member of the collection.
+        internal InlineCollection(IAvaloniaObject owner, bool isOwnerParent)
+            : base(owner, isOwnerParent)
+        {
+        }
+
+        #endregion Constructors
+
+        //-------------------------------------------------------------------
+        //
+        //  Public Methods
+        //
+        //-------------------------------------------------------------------
+
+        #region Public Methods
+
+        /// <summary>
+        /// Implementation of Add method from IList
+        /// </summary>
+        internal override int OnAdd(object value)
+        {
+            int index;
+
+            string text = value as string;
+
+            if (text != null)
+            {
+                index = AddText(text, true /* returnIndex */);
+            }
+            else
+            {
+                this.TextContainer.BeginChange();
+                try
+                {
+                    if (value is IControl uiElement)
+                    {
+                        index = AddUIElement(uiElement, true /* returnIndex */);
+                    }
+                    else
+                    {
+                        index = base.OnAdd(value);
+                    }
+                }
+                finally
+                {
+                    this.TextContainer.EndChange();
+                }
+            }
+            return index;
+        }
+
+        /// <summary>
+        /// Adds an implicit Run element with a given text
+        /// </summary>
+        /// <param name="text">
+        /// Text set as a Text property for implicit Run.
+        /// </param>
+        public void Add(string text)
+        {
+            AddText(text, false /* returnIndex */);
+        }
+
+        /// <summary>
+        /// Adds an implicit InlineUIContainer with a given UIElement in it.
+        /// </summary>
+        /// <param name="uiElement">
+        /// UIElement set as a Child property for the implicit InlineUIContainer.
+        /// </param>
+        public void Add(IControl uiElement)
```

Comment:
Previously `IAddChild`

Reply:
I changed the param to IControl (for now) here because TextElement's extend from StyledElement as far as I know.
This method however is meant for adding elements that should be auto-wrapped, such as images, etc. I think IControl might be a good target interface for that.